qualtran/bloqs/factoring/mod_arithmetic.py

```diff
@@ -0,0 +1,197 @@
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585) Fig. 6., 8.
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('p', bitsize=self.bitsize)])
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2 * self.bitsize, Toffoli())}
+
+    def short_name(self) -> str:
+        return 'x = -x mod p'
+
+
+@frozen
+class ModDbl(Bloq):
+    r"""An n-bit modular doubling gate.
+
+    Implements $U|x\rangle|p\rangle \rightarrow |2 * x mod p\rangle|p\rangle$ using $2n Toffoli$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        p: A bitsize-sized input register (register p above).
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585) Fig. 6., 8.
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature([Register('x', bitsize=self.bitsize), Register('p', bitsize=self.bitsize)])
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        return {(2 * self.bitsize, Toffoli())}
+
+    def short_name(self) -> str:
+        return 'x = 2 * x mod p'
+
+
+@frozen
+class ModMult(Bloq):
+    r"""An n-bit modular multiplication gate.
+
+    Implements $U|x\rangle|y\rangle|p\rangle|0\rangle \rightarrow |x\rangle|y\rangle|p\rangle|x * y mod p\rangle$ using $2.25n^2 + 9n Toffoli$ gates.
+
+    Args:
+        bitsize: Number of bits used to represent each integer.
+
+    Registers:
+        x: A bitsize-sized input register (register x above).
+        y: A bitsize-sized input register (register y above).
+        p: A bitsize-sized input register (register p above).
+        out: A bitsize-sized input register holding the output of the modular multiplication.
+
+    References:
+        [How to compute a 256-bit elliptic curve private key with only 50 million Toffoli gates](https://arxiv.org/abs/2306.08585) Fig. 6., 8.
+    """
+
+    bitsize: Union[int, sympy.Expr]
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            [
+                Register('x', bitsize=self.bitsize),
+                Register('y', bitsize=self.bitsize),
+                Register('p', bitsize=self.bitsize),
+                Register('out', bitsize=self.bitsize),
+            ]
+        )
+
+    def bloq_counts(self, ssa: Optional['SympySymbolAllocator'] = None) -> Set['BloqCountT']:
+        # Upper bound is 2.25 * (n ** 2) + 9n. We truncate to 2 so that the return value is always an integer.
```

Comment:
we should probably support floats here. Can you remove the truncation and open an issue to track the typing change?

Reply:
How can a Toffoli count be a float?

Reply:
idk it can be an average or a bound (per above)